### PR TITLE
Fix exception handling

### PR DESCRIPTION
**Pull request by Arguggi**

### CODE REVIEW

File: src/Control/Concurrent/CachedIO.hs [hunk 1]
```diff
@@ -3,69 +3,76 @@ module Control.Concurrent.CachedIO (
...
-          return $ refreshCache now cachedT
+          | isCacheStillFresh lastUpdated now -> return (return value)
+        -- There's data in the cache, but it's stale. Update the cache state
+        -- to prevent a second thread from also executing the action. The second
+        -- thread will get the stale data instead.
+          | otherwise -> do
+            writeTVar cachedT (Updating lastUpdated value)
+            return $ refreshCache (Fresh lastUpdated value) cachedT
+        -- Another thread is already updating the cache, just return the stale value
+        Updating _ value -> return (return value)
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: Come to think of it I suppose `Updating UTCTime a` should simplify to `Updating a`

File: src/Control/Concurrent/CachedIO.hs [hunk 2]
```diff
@@ -3,69 +3,76 @@ module Control.Concurrent.CachedIO (
...
-        (_, Right value) -> do
-          writeTVar cachedT (now, Right value)
-          return $ refreshCache now cachedT
+          | isCacheStillFresh lastUpdated now -> return (return value)
+        -- There's data in the cache, but it's stale. Update the cache state
+        -- to prevent a second thread from also executing the action. The second
+        -- thread will get the stale data instead.
+          | otherwise -> do
+            writeTVar cachedT (Updating lastUpdated value)
+            return $ refreshCache (Fresh lastUpdated value) cachedT
```
Comment (MHova) [inline src/Control/Concurrent/CachedIO.hs]: :neckbeard: capture `oldState@(Fresh lastUpdated value)` instead?

File: src/Control/Concurrent/CachedIO.hs (hunk 1 above)
Comment (Arguggi) [reply src/Control/Concurrent/CachedIO.hs]: Removed `UTCTime` from `Updating`.

File: src/Control/Concurrent/CachedIO.hs (hunk 2 above)
Comment (Arguggi) [reply src/Control/Concurrent/CachedIO.hs]: Captured

Comment (Arguggi) [reply src/Control/Concurrent/CachedIO.hs]: I changed all oldState  to previousState

**Comment by MHova** [comment]
👍 did you want to open a PR to sumall:master?

**Pull request merged**